libc/dirent/opendir.c

```diff
@@ -0,0 +1,398 @@
+	TEST_ASSERT_EQUAL_INT(EACCES, errno);
+	TEST_ASSERT_NULL(dirPtr);
+
+	/* Try to read from available directory inside locked directory */
+	errno = 0;
+	dirPtr = opendir(readable);
+	TEST_ASSERT_EQUAL_INT(EACCES, errno);
+	TEST_ASSERT_NULL(dirPtr);
+
+	/* No execute permission */
+	chmod(unreadable, 0600);
+	errno = 0;
+	dirPtr = opendir(readable);
+	TEST_ASSERT_EQUAL_INT(EACCES, errno);
+	TEST_ASSERT_NULL(dirPtr);
+
+	/* No read permission */
+	chmod(unreadable, 0300);
+	errno = 0;
+	dirPtr = opendir(unreadable);
+	TEST_ASSERT_EQUAL_INT(EACCES, errno);
+	TEST_ASSERT_NULL(dirPtr);
+
+
+	errno = 0;
+	chmod(unreadable, 0700);
+	rmdir(readable);
+	chmod(unreadable, 0000);
+}
+
+
+TEST(dirent_opendir, wrong_directory_name)
+{
+	errno = 0;
+	DIR *dirPtr = opendir(MAIN_DIR "/not_existing_directory");
+	TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+	TEST_ASSERT_NULL(dirPtr);
+
+	errno = 0;
+	dirPtr = opendir("");
+	TEST_ASSERT_NULL(dirPtr);
+	TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+}
+
+
+TEST(dirent_opendir, not_a_directory)
+{
+	errno = 0;
+	DIR *dirPtr = opendir(MAIN_DIR "/notadir.txt");
+	TEST_ASSERT_EQUAL_INT(ENOTDIR, errno);
+	TEST_ASSERT_NULL(dirPtr);
+}
+
+
+TEST(dirent_opendir, creating_dirs_in_closed_and_open_directories)
+{
+	/* Create dir in closed directory */
+	DIR *dirs[4];
+	test_mkdir_asserted(MAIN_DIR "/formerDir", 0777);
+
+	/* Create dir in opened directory, then close opened one */
+	DIR *dp = opendir(MAIN_DIR);
+	TEST_ASSERT_NOT_NULL(dirs[0] = opendir(MAIN_DIR "/formerDir"));
+
+	test_mkdir_asserted(MAIN_DIR "/latterDir", 0777);
+
+	closedir(dp);
+
+	/* Assure that both dirs can be opened without problems */
+	TEST_ASSERT_NOT_NULL(dirs[1] = opendir(MAIN_DIR "/formerDir"));
+	TEST_ASSERT_NOT_NULL(dirs[2] = opendir(MAIN_DIR "/latterDir"));
+	dp = opendir(MAIN_DIR);
+	closedir(dp);
+	test_mkdir_asserted(MAIN_DIR "/evenLatterDir", 0777);
+	TEST_ASSERT_NOT_NULL(dirs[3] = opendir(MAIN_DIR "/evenLatterDir"));
+
+	for (int i = 0; i < 4; ++i)
+		closedir(dirs[i]);
+
+	rmdir(MAIN_DIR "/formerDir");
+	rmdir(MAIN_DIR "/latterDir");
+	rmdir(MAIN_DIR "/evenLatterDir");
+}
+
+
+TEST(dirent_opendir, open_too_many_directories)
+{
+#ifdef OPEN_MAX
+	int dir_amount = OPEN_MAX + 100;
+	errno = 0;
+
+	TEST_ASSERT_EQUAL_INT(-1, test_create_directories(dir_amount));
+	TEST_ASSERT_EQUAL_INT(EMFILE, errno);
+#else
+	TEST_IGNORE_MESSAGE("OPEN_MAX not defined");
+#endif
+}
+
+
+TEST(dirent_opendir, open_small_enough_number_of_directories)
+{
+	errno = 0;
+
+	TEST_ASSERT_EQUAL_INT(0, test_create_directories(20));
+}
+
+
+TEST(dirent_opendir, open_same_dir_multiple_times)
+{
+	DIR *dp1, *dp2, *dp3;
+
+	dp1 = opendir(MAIN_DIR);
+	dp2 = opendir(MAIN_DIR);
+	dp3 = opendir(MAIN_DIR);
+
+	TEST_ASSERT_NOT_EQUAL(dp1, dp2);
+	TEST_ASSERT_NOT_EQUAL(dp2, dp3);
+	TEST_ASSERT_NOT_EQUAL(dp1, dp3);
+
+	closedir(dp1);
+	closedir(dp2);
+	closedir(dp3);
+}
+
+
+TEST(dirent_opendir, symlink_loop)
+{
+#ifndef SYMLOOP_MAX
+#define SYMLOOP_MAX 8
+#define SYMLOOP_NOT_DEFINED
+#endif
```

Comment:
```suggestion
#if !defined(SYMLOOP_MAX) && !defined(_SC_SYMLOOP_MAX)
TEST_IGNORE_MESSAGE("Symloops not supported");
#elif !defined(SYMLOOP_MAX) && defined(_SC_SYMLOOP_MAX)
#define SYMLOOP_MAX sysconf(_SC_SYMLOOP_MAX)
#endif
```

Reply:
As we discussed I'd like to do sth similar to the @adamdebek proposal:

```C
int symloopMax;
#if defined(SYMLOOP_MAX)
symloopMax = SYMLOOP_MAX;
#elif (sysconf(_SC_SYMLOOP_MAX) != -1)
symloopMax = sysconf(_SC_SYMLOOP_MAX)
#else
TEST_IGNORE_MESSAGE("Neither SYMLOOP_MAX nor sysconf(_SC_SYMLOOP_MAX) is defined");
#endif
```